tests/Unit/CssInlinerTest.php

```diff
@@ -2422,160 +2094,6 @@ function (array $dataset) {
-                $datasets[$leftDatasetName . ' & ' . $rightDatasetName]
-                    = \array_merge($leftDataset, $rightDataset);
-                if ($doneFirstLeft) {
-                    // Not all combinations are required,
-                    // just all of 'right' with one of 'left' and all of 'left' with one of 'right'.
-                    break;
-                }
-            }
-            $doneFirstLeft = true;
-        }
-        return $datasets;
-    }
-
-    /**
-     * @return string[][]
-     */
-    public function documentTypeAndSelfClosingTagDataProvider()
-    {
-        return self::joinDatasets($this->documentTypeDataProvider(), $this->selfClosingTagDataProvider());
-    }
-
-    /**
-     * @test
-     *
-     * @param string $documentType
-     * @param string $htmlWithNonXmlSelfClosingTags
-     * @param string $htmlWithXmlSelfClosingTags
-     *
-     * @dataProvider documentTypeAndSelfClosingTagDataProvider
-     */
-    public function renderConvertsXmlSelfClosingTagsToNonXmlSelfClosingTag(
-        $documentType,
-        $htmlWithNonXmlSelfClosingTags,
-        $htmlWithXmlSelfClosingTags
-    ) {
-        $subject = $this->buildDebugSubject(
-            $documentType . '<html><body>' . $htmlWithXmlSelfClosingTags . '</body></html>'
-        );
-
-        $result = $subject->render();
-
-        self::assertContains('<body>' . $htmlWithNonXmlSelfClosingTags . '</body>', $result);
-    }
-
-    /**
-     * @test
-     *
-     * @param string $documentType
-     * @param string $htmlWithNonXmlSelfClosingTags
-     *
-     * @dataProvider documentTypeAndSelfClosingTagDataProvider
-     */
-    public function renderKeepsNonXmlSelfClosingTags($documentType, $htmlWithNonXmlSelfClosingTags)
-    {
-        $subject = $this->buildDebugSubject(
-            $documentType . '<html><body>' . $htmlWithNonXmlSelfClosingTags . '</body></html>'
-        );
-
-        $result = $subject->render();
-
-        self::assertContains('<body>' . $htmlWithNonXmlSelfClosingTags . '</body>', $result);
-    }
-
-    /**
-     * @test
-     *
-     * @param string $htmlWithNonXmlSelfClosingTags
-     * @param string $tagName
-     *
-     * @dataProvider nonXmlSelfClosingTagDataProvider
-     */
-    public function renderNotAddsClosingTagForSelfClosingTags($htmlWithNonXmlSelfClosingTags, $tagName)
-    {
-        $subject = $this->buildDebugSubject(
-            '<html><body>' . $htmlWithNonXmlSelfClosingTags . '</body></html>'
-        );
-
-        $result = $subject->render();
-
-        self::assertNotContains('</' . $tagName, $result);
-    }
-
-    /**
-     * @test
-     */
-    public function renderAutomaticallyClosesUnclosedTag()
-    {
-        $subject = $this->buildDebugSubject('<html><body><p></body></html>');
-
-        $result = $subject->render();
-
-        self::assertContains('<body><p></p></body>', $result);
-    }
-
-    /**
-     * @test
-     */
-    public function renderReturnsCompleteHtmlDocument()
```

Comment:
We don't seem to have an equivalent test in `AbstractHtmlProcessorTest` so should this be moved there rather than removed?

Reply:
This is covered by several tests that test several smaller aspects of this:
- `reformatsHtml`
- `addsMissingHtml5DocumentType`

Reply:
...and `addsMissingContentTypeMetaTag` - I see.